fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -0,0 +1,356 @@
+ */
+@API(API.Status.EXPERIMENTAL)
+public class LucenePartitioner {
+
+    public static final int PARTITION_META_SUBSPACE = 0;
+    public static final int PARTITION_DATA_SUBSPACE = 1;
+    private final IndexMaintainerState state;
+    private final boolean partitioningEnabled;
+    private final String partitionTimestampFieldName;
+    private final Map<Tuple, List<LucenePartitionInfoProto.LucenePartitionInfo>> partitionMetaInfoCache = new HashMap<>();
+
+    public LucenePartitioner(@Nonnull IndexMaintainerState state) {
+        this.state = state;
+        partitionTimestampFieldName = state.index.getOption(com.apple.foundationdb.record.metadata.IndexOptions.TEXT_DOCUMENT_PARTITION_TIMESTAMP);
+        this.partitioningEnabled = partitionTimestampFieldName != null;
+    }
+
+    /**
+     * return the partition ID on which to run a query, given a grouping key.
+     * For now, the most recent partition is returned.
+     *
+     * @param grouping grouping key
+     * @return partition id, or null if partitioning isn't enabled
+     */
+    public Integer selectQueryPartitionId(Tuple grouping) {
+        Integer partitionId = null;
+        if (isPartitioningEnabled()) {
+            partitionId = 0;
+            List<LucenePartitionInfoProto.LucenePartitionInfo> partitions = getSortedPartitionMetadata(grouping);
+            if (partitions != null && !partitions.isEmpty()) {
+                partitionId = partitions.get(partitions.size() - 1).getId();
+            }
+        }
+        return partitionId;
+    }
+
+    /**
+     * return the partition to which a document with a given timestamp should be assigned.
+     * <b>Note:</b> {@link #loadPartitioningMetadata(Set)} should be called
+     * before the first call to this method.
+     * If no partition currently exists, a new one is created.
+     *
+     * @param grouping document grouping key
+     * @param timestamp document timestamp
+     * @return assigned partition metadata
+     */
+    public LucenePartitionInfoProto.LucenePartitionInfo assignWritePartition(Tuple grouping, long timestamp) {
+        List<LucenePartitionInfoProto.LucenePartitionInfo> partitions = partitionMetaInfoCache.get(grouping);
+        if (partitions == null) {
+            // if loadPartitioningMetadata(...) was called earlier, then this should never happen
+            throw new RecordCoreException("Internal Error");
+        }
+
+        if (partitions.isEmpty()) {
+            // very first partition
+            partitions.add(LucenePartitionInfoProto.LucenePartitionInfo.newBuilder()
+                    .setId(0)
+                    .setCount(0)
+                    .setFrom(timestamp)
+                    .setTo(timestamp)
+                    .build());
+            return partitions.get(0);
+        }
+
+        return selectWritePartition(partitions, timestamp);
+        // TO DO: check and split the assigned partition if it's too big
```

Comment:
Right. I didn't mean we would do the split here, but rather "queue" a split somewhere. Or did you mean that checking partitions for splitting happens independently and continuously via some groomer?

Reply:
I think it would be via some sort of client-provided queue.
It may also make sense to combine this and the merge into a single `IndexOperation` to maintain the index, and the `LuceneIndexMaintainer` will determine whether anything needs to be moved between partitions, and whether merges need to happen.